mail_embed_image/models/ir_mail_server.py

```diff
@@ -0,0 +1,130 @@
+
+
+logger = logging.getLogger(__name__)
+
+
+class IrMailServer(models.Model):
+    _inherit = 'ir.mail_server'
+
+    @contextmanager
+    def fetch_image(self, path):
+        public_user = self.env.ref('base.public_user')
+        session_store = root_wsgi.session_store
+        session = session_store.new()
+        session.update({
+            'db': threading.current_thread().dbname,
+            'login': public_user.login,
+            'uid': public_user.id,
+            'context': self.env.context,
+            })
+        werkzeug_env = EnvironBuilder(path).get_environ()
+        werkzeug_request = WerkzeugRequest(werkzeug_env)
+        werkzeug_request.session = session
+        # construct an odoo request with this werkzeug request.
+        request = http.HttpRequest(werkzeug_request)
+        with request:
+            request._env = self.env(user=public_user)
+            endpoint, arguments = http.routing_map(
+                self.env.registry._init_modules,
+                False,
+                self.env['ir.http']._get_converters()
+            ).bind_to_environ(
+                werkzeug_env).match(return_rule=False,)
+            yield endpoint, arguments
+
+    def build_email(
+            self,
+            email_from,
+            email_to,
+            subject,
+            body,
+            email_cc=None,
+            email_bcc=None,
+            reply_to=None,
+            attachments=None,
+            message_id=None,
+            references=None,
+            object_id=None,
+            subtype='plain',
+            headers=None,
+            body_alternative=None,
+            subtype_alternative='plain'):
+        result = super(IrMailServer, self).build_email(
+            email_from=email_from,
+            email_to=email_to,
+            subject=subject,
+            body=body,
+            email_cc=email_cc,
+            email_bcc=email_bcc,
+            reply_to=reply_to,
+            attachments=attachments,
+            message_id=message_id,
+            references=references,
+            object_id=object_id,
+            subtype=subtype,
+            headers=headers,
+            body_alternative=body_alternative,
+            subtype_alternative=subtype_alternative,
+        )
+        return self._build_email_replace_img_src(result)
+
+    def _build_email_replace_img_src(self, email):
+        """ Given a message, find it's img tags and if they
+        are URLs, replace them with cids.
+        """
+        for part in email.walk():
+            if part.get_content_type() == 'text/html':
+                body = part.get_payload(decode=True)
+                if not body or body == '\n':
+                    continue
+                root = self.process_img_body(fromstring(body), email)
+                # encodestring will put a newline every 74 char
+                part.set_payload(encodestring(tostring(root)))
+        return email
+
+    def process_img_body(self, root, email):
```

Comment:
this should be private too. It would make sense to prefix this function too with `_build_email` as the above does, indicating they belong together